### PR TITLE
More Attack Animations

### PR DESCRIPTION
**Pull request by RemieRichards**
- Melee weapons used in combat now appear over the victim for 3/10ths of a second.
- Fixes SNPCs thinking a ranged attack was a melee attack, causing this new animation to trigger for guns, rofl, sawu accidentally reused a proc to avoid writing 2 lines... yeah, haha.
- `flick_overlay()` now uses `spawn()` instead of `sleep()` so that `animate()` and other effects can be called on the image while it is displayed.

I feel this is appropriate as when you see combat you already see "X hits Y with the Z" so showing a picture of Z is no more information, it's just a better way of conveying the same information.

combat2

### CODE REVIEW

**Comment by optimumtact** [comment]
Yeah we fucking talking now

**Comment by xxalpha** [comment]
This could get confusing. Maybe if they were a different color (even though that would get really expensive).

**Comment by RemieRichards** [comment]
@xxalpha it wouldn't get expensive at all, this is an image not an icon so I can use .color.
but I don't feel colouring the image is appropriate, that'd just get MORE confusing.

**Comment by freerealestate** [comment]
I think I'd prefer if the weapon was shown in the chat window beside its
name (or instead of its name), rather than this. This is a little too meta
(for lack of a better term), like Goofball's proposed pop-up damage numbers
were.

**Comment by RemieRichards** [comment]
@freerealestate if we had the time for actual animations and such, you'd see the weapon fly into/over/onto the person... this is just that but a more reasonable level of work, this requires no new animations ever be added for any new weapons added either.

**Comment by phil235** [comment]
I like the concept but witnesses may confuse the weapon shown for a real weapon on the ground. Something needs to be added to the image to not be confused with the an actual weapon.

**Comment by RemieRichards** [comment]
I think the fact it exists for 1/2 a second is enough to be able to tell the difference.
AND the fact it's above the person, not below them.

**Comment by xxalpha** [comment]
It's too meta as freerealestate said, but besides that, it does need something to differenciate it visually from a dropped item. Personally the short time it's on the screen is not enough, that looks the same as the victim dropping a weapon and the attacker picking it up.

We need configs to show/hide these things (same bag as the progress bars). 2meta4me

**Comment by RemieRichards** [comment]
Again, how is it meta, if this were IRL or we had better animations, the weapon would be over the person, same as it is now.

**Comment by kingofkosmos** [comment]
Is it possible to add transparency?

**Comment by xxalpha** [comment]
It's a UI element being shown in the world (to everybody). That's how I see it.

**Comment by RemieRichards** [comment]
@kingofkosmos yes, but, I don't feel it would look better, it would probably just look confusing.

**Comment by Dorsisdwarf** [comment]
I dunno about gameyness, and I like the idea, but it looks *awful*

**Comment by AnturK** [comment]
I'd add pulse effect and made the weapon scaled down a bit maybe.

**Comment by RemieRichards** [comment]
@AnturK by pulse I assume you mean have it flash like, white or something?
and I'm deliberately avoiding scaling it because byond's scaling is still not NN (it recently became NN for the map, but for nothing else) so it looks horrific.

**Comment by AnturK** [comment]
Yeah something like that. So it's at once visible it's UI/Effect not an item.

**Comment by Dorsisdwarf** [comment]
Maybe make the image over them spin/move from the attacker onto the target and back?

**Comment by RemieRichards** [comment]
@Dorsisdwarf I'm sorry, how is that not objectively more hideous.

**Comment by Dorsisdwarf** [comment]
It's better than a ghost item appearing for a noticable amount of time and vanishing. Like I said, your gif example in the top looks dreadful. 

Sorry.

**Comment by RemieRichards** [comment]
Pulsing a colour or Fading in and out are off the table, it's an image in the overlays list, meaning it does not visibly update to show these changes in colour/alpha transparency.

**Comment by Kelenius** [comment]
Maybe the weapon could be shown between the attacker and the target?

**Comment by RemieRichards** [comment]
@Kelenius I fear that would lead to "Who attacked Who?" especially if A is attacking B AND B is attacking back, the two weapons would both overlay in the middle.

**Comment by CosmicScientist** [comment]
I'm... not sure. I don't know if I would have ever been concerned about this since different weapons have different sounds. E-sword: cry. Slash: probably easy to handle. Punch: you're taking the mickey. I would still be in favour if we didn't have big and bold text for when the damage is being dealt to you though... but this does make information more available... 

Maybe if it used different sprites? Say just the head of the stun baton (still shows if it's on or off) or if it used just the head of the cleaver? I mean it covers up a lot and as others said it could be a bit of distraction looking like a dropped weapon until you realise it's over the body, not under.

Actually I do think this is good to see for when that one person is attacking that other person in the corner of the room you can't tell what's going on because you can't read the side text in half a second flat.

Do guns make a kapow bullet flash? I would ask for it to display over the limb it hit but when you're facing left, right or even backwards that'd be hard to ask for and I suppose it'd ruin the fun chaos of "Please fall over please fall over!" clicking brings.

**Comment by WJohn** [comment]
I dunno, I find this looks really ugly.

Maybe make it so the inhand is shifted forward on top of the attacker's shift? It'd still tell you what weapon was used but it wouldn't be so blatantly obvious.

**Comment by RemieRichards** [comment]
@CosmicScientist the goal is to reuse the item sprites, to have to produce a new sprite for every attack animation immediately devalues the system by a googolplex %, it becomes worthless.

@WJohn... I have no idea what you're saying

**Comment by WJohn** [comment]
If you attack someone with a weapon, the inhand sprite of that weapon should shift forward in addition to you also shifting from attacking.

**Comment by RemieRichards** [comment]
@WJohn it already does...

unless you mean split the weapon off of the person and then shift the inhand forward... you of all people should know the inhand sprites are NOT built for that, with holes cut out of them all over the place.

**Comment by CosmicScientist** [comment]
How does it devalue the system? I don't actually know.

**Comment by WJohn** [comment]
Yeah I know they have holes built into them but I'm really not keen on there being the giant on floor sprite of the item appearing on the victim.

**Comment by RemieRichards** [comment]
@CosmicScientist it adds extra work, this system adds no new work.
With the system you're proposing, that's 1-2 new sprites needing to be drawn for EVERY ITEM IN THE GAME PRESENT, + 1-2 more per every new item added.

With the current system you add nothing extra or new and it just *works*

**Comment by RemieRichards** [comment]
@WJohn I can't believe that you, one of our art maintainers who's job it is to keep the game's look and feel right, are telling me that a sprite with ugly holes in it looks better than a perfectly good, perfectly solid (No holes!) sprite being reused.

**Comment by CosmicScientist** [comment]
Ah okay then.

Are you able to flip the image perhaps? I know some items look almost the same both ways such as a toolbox but nothing serious comes to mind as being symmetrical.

**Comment by RemieRichards** [comment]
@CosmicScientist without testing, I'd say "maybe", but that would most likely require swapping /image for /icon (which incurs extra lag) and then it might not even update properly in the overlays.

But my main issue with that is the fact that some items do actually have directional sprites other than south, and those directional sprites are just the image flipped or rotated, so it would still look the same as "an item on the floor"

**Comment by CosmicScientist** [comment]
Alright then. It's a nice idea. I like it for being able to see what's happening to other people since it's easy to focus on the person hurting you in the game thanks to purposeful sprites but it's not as easy to look at something happening between two other combatants. And it's also nice feedback that audio/text can't currently achieve.

I also disagree with above criticism that says this is like goof's damage numbers, this is just helping us see information we already get without changing the feel of the game.

**Comment by Steelpoint** [comment]
I thought what was being shown was a bug for a moment.

I like the concept but I dislike the implementation, it looks like a item just appears and disappears instantly on someone.

It just does not look right nor good.

**Comment by Kelenius** [comment]
Maybe the big sprite could just slide from the attacker to the attacked instead then?

**Comment by kingofkosmos** [comment]
If it's possible to add transparency, would it be possible to have it fade out?

**Comment by RemieRichards** [comment]
@kingofkosmos 

>Pulsing a colour or Fading in and out are off the table

**Comment by SconesC** [comment]
This looks... Really bad.

Like an enlarged sprite is just stuttering on top of someone. Can it be a pref? It's just going to be annoying otherwise

**Comment by Tokiko1** [comment]
I find it rather confusing. I feel like I would not be able to reliably tell the difference between a dropped and quickly picked up weapon and a person being hit unless I focus my eyes on the mob to confirm that the sprite did in fact appear on top of the mob(or read the logs). Certainly not always possible for me in a chaotic combat situation.

**Comment by SconesC** [comment]
If there is a problem with weapons being easily identifiable in combat the solution is to add more unique weapon sounds, not a giant flashing icon

**Comment by Steelpoint** [comment]
Maybe even look at firing animations for when your firing a gun?

**Comment by Dorsisdwarf** [comment]
I also don't see how this helps identify who is attacking who more than the current attackimations do.

**Comment by Iamgoofball** [comment]
shoulda accepted damage numbers

:+1:

**Comment by Steelpoint** [comment]
It would be amusing to have toggable features like damage numbers appearing above people you attack, or anyone attacks.

**Comment by Iamgoofball** [comment]
@Steelpoint you saw the PR I made for that right

**Comment by palpatine213** [comment]
:-1: the idea is interesting, but the implementation is poor. This is going to be annoying if someone disarms you, then starts attacking you with a weapon with a large sprite. You're not going to be able to see under your character to pick up what you dropped.
Also, does the image copy the item exactly in size? 
I'm seeing the possible horror of a botanist attacking someone with 100 potency death nettles and not being able to see anything nearby when the sprite pops up.

**Comment by MrPerson** [comment]
You can do flipping or even arbitrary rotations with images. The transform var is available.

We should make flick_overlay() spawn instead of sleep. Then you could call animate() on the image after calling flick_overlay() and it would work fine. Then you could do stuff like a slide or a gradual rise up.

**Comment by RemieRichards** [comment]
@palpatine213 Botanist plants do not have bigger icons. they are simply rendered bigger because the matrix used for their transform is scaled up based on their potency/yield, this code simply takes the sprite and icon file from the object, it does not inherit any of the scaling.

@MrPerson completely forgot images had the transform var, however I don't think flipping would actually make a difference as mentioned above about some items actually having directional sprites that are simply rotations of the south one.
fixing ```flick_overlay()``` is on the cards too, I'll look into that later.
Maybe try the pulse/fade effects again

**Comment by SconesC** [comment]
@RemieRichards If you are still going ahead with it, PLEASE make it a pref

**Comment by freerealestate** [comment]
The thing about making it a pref is that it still conveys an advantage to those who enable it, so those who don't like it are forced to put up with it or be at a disadvantage. Not a big deal, but I thought I'd mention it.

**Comment by SconesC** [comment]
"Advantage"

Yeah lol how about no

**Comment by Steelpoint** [comment]
One alternative I can think of is to reduce the icon size to roughly half of what it is, move it upwards towards the head and change the outline of the object to reflect the dominate type of damage its dealing.

**Comment by Killbasa** [comment]
Do we really need something like this? I find it easy enough to tell what is happening from the sounds.

**Comment by freerealestate** [comment]
Uh, this pull wouldn't be here if it didn't provide an advantage.

**Comment by SconesC** [comment]
A giant flashing icon for some info I could already discern from sounds/chat honestly just sounds like a hindrance

**Comment by freerealestate** [comment]
The sounds aren't unique though, and looking at the chat means having to
take your eyes off of combat.

**Comment by Alek2ander** [comment]
A tiny icon in the chat would do better.

**Comment by xxalpha** [comment]
>A tiny icon in the chat would do better.

I like this idea.

**Comment by SconesC** [comment]
Still sounds annoying and cluttery 

Make it a pref

**Comment by Razharas** [comment]
On one hand its useful, on the other hand i dont enjoy how it looks like

**Comment by WJohn** [comment]
Tiny icon in chat sounds good, we already do this when examining someone and it helps a bunch. Having examine be full text would be a real pain in the ass.

**Comment by SconesC** [comment]
I don't want little icons on everything in chat, it sounds like clutter and would look EXTREMELY annoying

**Comment by MrPerson** [comment]
The idea is to give the player all the information they need in the place they're actually looking; the center of the screen.

I think having it fade and drift upwards a bit (like in goofball's damage number pr) would make it obviously not an object falling on the ground.

**Comment by Rinoka** [comment]
Upsides of the system: You can look at a brutal beating and tell what's happening without any sort of struggle. I appreciate this personally (since I am particularly slow sometimes) and it makes some situations more sensical (like large scale fights, like the sort you see post-round on the shuttle).

Downsides of the system: It's definitely a boon to new players (oh no) with the trade off that if you're the one being robusted, you'd have to alt+click yourself to try and grab anything, and for the sake of robustness, that does not always work well.

If you want to do this, I'd say find a way to cleanly rescale it. Since it's a useful visual tool, too, maybe just make it clear from the overlay where the object is coming from? IE a slight animation from the attacker towards the center of the person's tile, establishing direction? If a group is browbeating the shit out of the clown, you want to know who's going at him with a baseball bat compared to who's going at him with a ball of twine.

**Comment by Fayrik** [comment]
First of all, .color should hopefully allow for transparency, and I think that's the first place to look for any changes here. You can tell it's not an icon on the ground, if it looks like a ghost of the item used.

Secondly, what I'd rather see is a generic "swoosh" effect, which would look much cooler.
Combine the too, with a very faint ghost item and a more noticeable swoosh effect, seems like it would be a more palatable change for everyone.

Personally I'd also prefer a less meta solution, but I don't have one in mind so I guess I can't judge too much.

**Comment by PKPenguin321** [comment]
looks pretty neat but really obstructive, maybe make it a bit transparent

**Comment by WJohn** [comment]
Oh yeah, maybe instead make it so we get a transparent object appearing over them, which then rises and fades away? That'd be much less jarring.

**Comment by SconesC** [comment]
Seriously do not force this on people

It's always going to look extremely obstructive no matter what you do to it

**Comment by Rinoka** [comment]
Also, for those situations where you are determining how best to beat the clown to death in a horrifying mob because they played with new slip mechanics and got a toolbox embedded in someone for their trouble, it's best to make sure this is clickthroughable, akin to darkness/alignment icons.

**Comment by Cheridan** [comment]
I like Steelpoint's idea of having the icon be at 50% size. Most of our item icons are oversized for the sake of clickability, shrinking them a bit might look nicer.

**Comment by Cebasiz** [comment]
i don't know, does anyone have any problems deciphering combat after they've played the game for a few hours? 

ages ago someone added the hit animation and i don't think anything else is necessary really. i don't like the way it looks and i don't think anybody needs more visual information cluttering the ground. 

it's still not as bad as goofnumbers though. my goodness.

**Comment by optimumtact** [comment]
It's about identifying at a glance what item is hitting who.

Make this a pref, but I want it so bad, I'm happy to give up some immersion for some powergame

Also thought, does this affect pens/other stealthy items?

**Comment by Lumipharon** [comment]
This is just confusing as shit. It looks like people are dropping or throwing items.

**Comment by optimumtact** [comment]
Are you serious? It looks nothing like that at all

At least come up with an argument that isn't weaksauce as shit

**Comment by Killbasa** [comment]
Yes, it does. 
I would also say that it provides no advantage whatsoever, i would even consider a hinderance. It is easy as shit to tell what is happening by looking at their sprite, if there is any blood, looking at the text that is showing up, and even the sounds of getting hit.

**Comment by palpatine213** [comment]
I honestly don't think that you should be able to instantly see exactly what someone is attacking another person with just at a glance at the corner of your screen when they are just at the edge of your view.
You should have to stop for a moment and analyze the situation by at least reading the text box to see what's going on.

**Comment by Rinoka** [comment]
Okay let me just

@Killbasa No. Just no. Items layer under mobs, even if they're dropped by that mob. Something layering over the mob would clearly be a sign of use, but if it's flickering and moves (as we've been suggesting), then it's actually relatively easy to tell.

@palpatine213 There really is zero harm in this, especially since some people are on the level where they understand everything and some people are not that great at it. What this is is giving people training wheels.

I think that if this is added, it should be a toggle defaulted to off, but it should be added. Adding this just evens the playing field between people who read faster than others, and I support that.

**Comment by Lumipharon** [comment]
It also means you can see 100% clearly via cameras (which otherwise don't let you examine or read damage notifications when say... cultist McGee is bludgeoning a guy to death with a tome.

This is also just super immersion breaking, and is clogging the screen with what's akin to MMO tier effects on people.

**Comment by Iamgoofball** [comment]
why do we need toggles for everything

**Comment by Iamgoofball** [comment]
>immersion breaking

oh boy here we go again

**Comment by Lumipharon** [comment]
Please explain why that's not a good argument rather then using obama tier discrediting.
It's literally giant items appearing over people when you poke them.

**Comment by Iamgoofball** [comment]
okay i'll cede on the toggle thing, just toggle it off then

while robust pros like me will enable it

**Comment by Rinoka** [comment]
And so many people have voted to keep the idea and change how the icons display. Bro.

It is immersion breaking but in the world of /tg/ validhunting and metagaming, I don't think all of us care, especially if it's toggleable.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

I actually don't care for this, at all, but I can understand the reasoning for it--I do tend to agree with the suggestion of adding in icons to the chat to show which weapon was used, but this is kinda...I dunno; it makes things feel really "gamey", but that's just me.

Interesting concept, none the less.

**Comment by Iamgoofball** [comment]
gameplay > realism

**Comment by SconesC** [comment]
but this isn't good gameplay goof its a massive visual obstruction and eyesore

**Comment by Iamgoofball** [comment]
then find a way to add it that isn't either of those things

**Comment by Rinoka** [comment]
@SconesC 

please actually pay attention to discussion, especially when it discusses not making it an eyesore and putting it on a toggle

**Comment by SconesC** [comment]
@Rinoka Step off, no reply from Remie about making it a toggle, and there is no way to not make it an eyesore.

**Comment by PKPenguin321** [comment]
could make it smaller and transparent

**Comment by Dorsisdwarf** [comment]
Make it appear in the chatbox.

**Comment by Lumipharon** [comment]
If you can't A: see what they're holding in their hand 
or B: be bothered glancing at the chat
then it would better to have distinctive sounds, since right now it's basically:
e-sword sound, saw sound, misc stabbing sound, misc thunking sound etc.

This ugly visual pollution is pretty terrible, even if it is toggleable.

**Comment by PKPenguin321** [comment]
if it is toggleable then who cares

**Comment by kingofkosmos** [comment]
How is this immersion breaking? In real life you would be able to tell with a glance what weapon is being used in an attack.

**Comment by Rinoka** [comment]
@SconesC and very few of us are supporting the idea without a toggle. No comment on eyesore, since it's not really valid to just declare it is/isn't until it's been messed with a bit more.

The only reason it's immersion breaking to some is because it's hard to visualize how it could be done more smoothly. I still think it's possible to make it work.

**Comment by freerealestate** [comment]
It's more the fact that this would add floating icons to the world that
everyone can see without an IC reason to explain them. This in a game that
tries to steer clear of gaming abstractions.

Adding a toggle would still mean choosing between using this or being at a
disadvantage.

**Comment by RemieRichards** [comment]
## THE EFFECT HAS CHANGED
combat2

I'm much more pleased with this one, it has agency and viscera too it, thanks to MrPerson for the 
sleep->spawn() idea and anyone who suggested a fading effect.

**Comment by RemieRichards** [comment]
The Travis issue is with ministation, not this PR.

**Comment by freerealestate** [comment]
Looks better. It's more apparent with the transparency that it's part of
the UI and not in the world.

Don't like how the icon rotates with the mob, but that's just nitpicking.

The icon could probably last a little longer.

**Comment by RemieRichards** [comment]
@freerealestate it lasts as long as the pixel shift of the attacking player does, so it lasts an entire "instance" of combat.

The icon rotation is a limitation of byond, I know one way of *fixing* it, but that would leave it on the turf meaning people could walk away from them... which I consider much worse!

**Comment by Alek2ander** [comment]
What if it's attached to the attacker instead of the target? You usually can't attack when prone, so this becomes less of an issue.

**Comment by RemieRichards** [comment]
@Alek2ander the victim can once again move, meaning you have a visual effect saying combat occurred in one place, and a bleeding runaway saying it happened where they're standing.

tl;dr, confusing city.

**Comment by Alek2ander** [comment]
The icon doesn't stay for very long, and that time can be shortened so by the time the victim is halfway off his tile it's already gone and not confusing anyone.

**Comment by Ricotez** [comment]
:+1: much better effect

**Comment by CosmicScientist** [comment]
Jesus Space Chroist I'm loving it even more now. :+1:

**Comment by Killbasa** [comment]
It looks better now, but i still don't think we need this.

**Comment by Nanotransistor** [comment]
Why not replace the item(the axe here) animation with a blood splash?

**Comment by RemieRichards** [comment]
@Nanotransistor that would be generic, and for a slightly different reason than this.
This is to show that combat is occuring visual (as would a blood splatter) but this also allows you to see what weapons are being used without having to know the tiny inhand sprites of the weapons or look in chat.

**Comment by Alek2ander** [comment]
Minor consideration: hallucinating people should see the overlay when fighting a fake attacker.

**Comment by Razharas** [comment]
Thats better
But can you make it unturnable when man is downed?
Otherwise someone might think its some special regime where you strike with handle, since it looks exactly like that

**Comment by PKPenguin321** [comment]
is there an icon for fists? just wondering

**Comment by Invisty** [comment]
I would say it looks OK currently. It's a significant improvement over having to avert one's eyes to read the red text. 

As a potential improvement, maybe have the animation occur between the two mobs, moving a few pixels from the attacker > towards the victim?

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -782,6 +782,23 @@ Sorry Giacom. Please don't be mad :(
...
+		if(r_hand)
+			I = image(r_hand.icon,A,r_hand.icon_state,A.layer+1)
+	if(I)
+		var/list/viewing = list()
+		for(var/mob/M in viewers(A))
+			if(M.client)
+				viewing |= M.client
+		flick_overlay(I,viewing,5)
+		I.pixel_z = 16 //lift it up...
+		animate(I, pixel_z = 0, alpha = 125, 3) //smash it down into them!
```
Comment (MrPerson) [inline code/modules/mob/living/living.dm]: You can just call animate() with a 3 like that???!!? wtf byond pls

Even if you can, I would probably make it time = 3 just for clarity

Comment (RemieRichards) [inline code/modules/mob/living/living.dm]: While most byond procs work with:
```proc(arg1 = "A",arg2="B")````
and also
```proc("A","B")```

I get that animate it a reaaaaal weird beast considering it has infinite (or close) arguments due to being able to put as many variables to shift through the animation.

I'll clarify it as the time.

**Comment by RemieRichards** [comment]
@Razharas again, I can't easily fix the rotation:
1. Either the rotation is fixed, and the mob is able to WALK AWAY FROM THE ANIMATION
or
2. The Animation rotates, but it follows the mob around correctly.

I opted for 2 as it makes the most sense visually, the rotation only looks wonky on some weapons, and until someone raises a genuine fix for both 1 and 2, I'm sticking with 2.

@Invisty I would, but I feel that would lead to confusion as to who's being attacked, and I doubt shifting it like, 2 pixels towards the attacker would give the effect you're asking for.

**Comment by duncathan** [comment]
This looks far nicer. :+1:

**Comment by WJohn** [comment]
I think I'm okay with this, although I think it being a bit more transparent would be nice.

**Comment by Lumipharon** [comment]
Did this really get merged despite all the opposition to it, without even a toggle?

**Comment by freerealestate** [comment]
It doesn't look very good.

**Comment by palpatine213** [comment]
Vote on sybil had the animations look bad as the dominant vote.
Also, tk attacks use the TK red circle thing as the visible attack, which looks really ugly.

**Comment by WJohn** [comment]
Yeah frankly this still needed a bit of tweaking, needs to be more transparent and very nearly fade away completely by the end of its animation.

**Comment by SconesC** [comment]
Looks pretty terrible

So many people asked for a pref

No pref

??????????????????????????????????

**Comment by Iamgoofball** [comment]
I dunno what you guys are going off about, it's really transparent and fades away pretty fast on live.

**Comment by freerealestate** [comment]
It just looks really bad. A lot of the rotations and positions of the items
don't look right.
